include/caffe/util/io.hpp

```diff
@@ -9,26 +11,43 @@
 #include "caffe/common.hpp"
 #include "caffe/proto/caffe.pb.h"
 
+#ifndef CAFFE_TMP_DIR_RETRIES
+#define CAFFE_TMP_DIR_RETRIES 100
+#endif
+
 namespace caffe {
 
 using ::google::protobuf::Message;
 using ::boost::filesystem::path;
 
-inline void MakeTempFilename(string* temp_filename) {
-  temp_filename->clear();
-  const path& model = boost::filesystem::temp_directory_path()
-    /"caffe_test.%%%%%%";
-  *temp_filename = boost::filesystem::unique_path(model).string();
-}
+static path temp_files_subpath;
+static uint64_t next_temp_file = 0;
 
 inline void MakeTempDir(string* temp_dirname) {
   temp_dirname->clear();
-  const path& model = boost::filesystem::temp_directory_path()
-    /"caffe_test.%%%%%%";
-  const path& dir = boost::filesystem::unique_path(model).string();
-  bool directoryCreated = boost::filesystem::create_directory(dir);
-  CHECK(directoryCreated);
-  *temp_dirname = dir.string();
+  const path& model =
+    boost::filesystem::temp_directory_path()/"caffe_test.%%%%-%%%%";
+  for ( int i = 0; i < CAFFE_TMP_DIR_RETRIES; i++ ) {
+    const path& dir = boost::filesystem::unique_path(model).string();
+    bool done = boost::filesystem::create_directory(dir);
+    if ( done ) {
+      *temp_dirname = dir.string();
+      return;
+    }
+  }
+  LOG(FATAL) << "Failed to create a temporary directory.";
+}
+
+inline void MakeTempFilename(string* temp_filename) {
+  temp_filename->clear();
+  if ( temp_files_subpath.empty() ) {
+    string path_string="";
+    MakeTempDir(&path_string);
+    temp_files_subpath = path_string;
+  }
+  std::ostringstream s;
+  s << std::setw(9) << std::setfill('0') << next_temp_file++;
+  *temp_filename = (temp_files_subpath/s.str()).string();
```

Comment:
@eelstork So I see this implementation follows the Apple guide "The best way to handle this (Create Temporary Files Correctly) is to create a safe temporary directory that only you can access, then write the files into that directory."

Looks good to me.